Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp

```diff
@@ -0,0 +1,536 @@
+ * call and will return after the thread has started
+ */
+void KafkaEventStreamDecoder::startCapture(bool startNow) {
+
+  // If we are not starting now, then we want to start at offsets corresponding
+  // to the start of the run
+  if (!startNow) {
+    auto runStream = m_broker->subscribe(m_runInfoTopic);
+    std::string rawMsgBuffer;
+    runStream->consumeMessage(&rawMsgBuffer);
+    if (rawMsgBuffer.empty()) {
+      throw std::runtime_error("KafkaEventStreamDecoder::initLocalCaches() - "
+                               "Empty message received from run info "
+                               "topic. Unable to continue");
+    }
+    auto runMsg = ISISStream::GetRunInfo(
+        reinterpret_cast<const uint8_t *>(rawMsgBuffer.c_str()));
+    auto eventOffset = runMsg->stream_offset();
+    m_eventStream = m_broker->subscribe(m_eventTopic, eventOffset);
+  } else {
+    m_eventStream = m_broker->subscribe(m_eventTopic);
+  }
+
+  m_runStream = m_broker->subscribe(m_runInfoTopic);
+  m_spDetStream = m_broker->subscribe(m_spDetTopic);
+
+  auto m_thread = std::thread([this]() { this->captureImpl(); });
+  m_thread.detach();
+}
+
+/**
+ * Stop capturing from the stream. This is a blocking call until the capturing
+ * function has completed
+ */
+void KafkaEventStreamDecoder::stopCapture() noexcept {
+  // This will interrupt the "event" loop
+  m_interrupt = true;
+  // Wait until the function has completed. The background thread
+  // will exit automatically
+  while (m_capturing) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  };
+}
+
+/**
+ * Check if there is data available to extract
+ * @return True if data has been accumulated so that extractData()
+ * can be called, false otherwise
+ */
+bool KafkaEventStreamDecoder::hasData() const noexcept {
+  std::lock_guard<std::mutex> lock(m_mutex);
+  return !m_localEvents.empty();
+}
+
+/**
+ * Check if a message has indicated that end of run has been reached
+ * @return  True if end of run has been reached
+ */
+bool KafkaEventStreamDecoder::hasReachedEndOfRun() noexcept {
+  // Notify the decoder that MonitorLiveData knows it has reached end of run
+  // and after giving it opportunity to interrupt, decoder can continue with
+  // messages of the next run
+  if (!m_extractedEndRunData)
+    return false;
+  if (m_endRun) {
+    std::lock_guard<std::mutex> runStatusLock(m_runStatusMutex);
+    m_runStatusSeen = true;
+    cvRunStatus.notify_one();
+  }
+  return m_endRun;
+}
+
+/**
+ * Check for an exception thrown by the background thread and rethrow
+ * it if necessary. If no error occurred swap the current internal buffer
+ * for a fresh one and return the old buffer.
+ * @return A pointer to the data collected since the last call to this
+ * method
+ */
+API::Workspace_sptr KafkaEventStreamDecoder::extractData() {
+  if (m_exception) {
+    throw * m_exception;
+  }
+
+  m_extractWaiting = true;
+  cv.notify_one();
```

Comment:
Nice use of condition variables. Does the static nature of the variable mean that we can then only have 1 listener running at a time?

(Admittedly I'm not sure if users would every have more 1 running...)